### DISCUSSION TITLE
QST/ API/BUG Inconsistency between getitem and .loc / other indexing methods:

### DISCUSSION THREAD

**Discussion by m-richards** [geopandas/geopandas#2726] [Q&A]
I've been looking at following up the active geometry column changes and looking at the removal of `__getitem__`. In updating the error messages in #2329 I've noticed that introduces another inconsistency (taken from `test_geodataframe.py::test_get_geometry_invalid`):
```python
gdf = gpd.read_file(gpd.datasets.get_path("nybb"))
gdf.columns
Index(['BoroCode', 'BoroName', 'Shape_Leng', 'Shape_Area', 'geometry'], dtype='object')
gdf['geom2'] = gdf.geometry
gdf2 = gdf[["BoroCode", "BoroName", "geom2"]]
gdf2.geometry
Traceback (most recent call last):
  File "C:\Users\Matt\.conda\envs\py39_pd14_gpd_dev\lib\code.py", line 90, in runcode
    exec(code, self.locals)
  File "<input>", line 1, in <module>
  File "C:\Users\Matt\.conda\envs\py39_pd14_gpd_dev\lib\site-packages\pandas\core\generic.py", line 5575, in __getattr__
    return object.__getattribute__(self, name)
  File "C:\Data\myDocuments\Code\geopandas\geopandas\geodataframe.py", line 233, in _get_geometry
    raise AttributeError(msg)
AttributeError: You are calling a geospatial method on the GeoDataFrame, but the active geometry column to use has not been set. 
There are columns with geometry data type (['geom2']), and you can either set one as the active geometry with df.set_geometry("name") or access the column as a GeoSeries (df["name"]) and call the method directly on it.
gdf3 = gdf.loc[:, ["BoroCode", "BoroName", "geom2"]]
gdf3.geometry
Traceback (most recent call last):
  File "C:\Users\Matt\.conda\envs\py39_pd14_gpd_dev\lib\code.py", line 90, in runcode
    exec(code, self.locals)
  File "<input>", line 1, in <module>
  File "C:\Users\Matt\.conda\envs\py39_pd14_gpd_dev\lib\site-packages\pandas\core\generic.py", line 5575, in __getattr__
    return object.__getattribute__(self, name)
  File "C:\Data\myDocuments\Code\geopandas\geopandas\geodataframe.py", line 233, in _get_geometry
    raise AttributeError(msg)
AttributeError: You are calling a geospatial method on the GeoDataFrame, but the active geometry column ('geometry') is not present. 
There are columns with geometry data type (['geom2']), and you can either set one as the active geometry with df.set_geometry("name") or access the column as a GeoSeries (df["name"]) and call the method directly on it.
```
specifically that the new getitem error refers explicitly to the active geometry column not being set, rather than the last active geometry column being missing. I'm not actually sure which behaviour is most clear to an end user. If as in this case the active geometry column was dropped, it perhaps seems more useful to know that there used to be a geometry column called geometry, rather than saying that the active geometry column is None, even if saying the geometry column is None is more technically correct.

Either way I think this is an inconsistency which should probably be resolved (and thereby letting `__getitem__` be removed without introducing inconsistency.

It's also worth noting that the behaviour of assigning the default geometry column name of "geometry" further confuses matters - as the case of `df = GeoDataFrame()` will say that "geometry" is missing, rather than saying no geometry column is set.

**Comment by jorisvandenbossche** [comment]
Yeah, I noticed that inconsistency recently as well. Thanks for the write up.

I also don't have a strong opinion on which of the two is correct. If you start from a GeoDataFrame with a geometry column set, and then drop this column in some way (eg by not selecting that column, as in the examples above), it's maybe indeed a bit more informative to say that the active geometry column is not present anymore.

> It's also worth noting that the behaviour of assigning the default geometry column name of "geometry" further confuses matters - as the case of `df = GeoDataFrame()` will say that "geometry" is missing, rather than saying no geometry column is set.

Yes, agreed. Especially now that we have better error messages for this, it might be a good time to reconsider that.

**Comment by jorisvandenbossche** [comment]
While looking at the above example, I was also wondering, if the geom2 and geometry names are switched (geom2 is active) and then drop geom2, should `.geometry` then raise the error? Currently we then return the non-active "geometry" column:


```
In [25]: gdf = gpd.read_file(gpd.datasets.get_path("nybb"))

In [26]: gdf['geom2'] = gdf.geometry

In [27]: gdf2 = gdf.set_geometry("geom2")

In [28]: gdf3 = gdf2.loc[:, ["BoroCode", "BoroName", "geometry"]]

In [29]: gdf3._geometry_column_name
Out[29]: 'geom2'

In [30]: gdf3.geometry
Out[30]: 
0    MULTIPOLYGON (((970217.022 145643.332, 970227....
1    MULTIPOLYGON (((1029606.077 156073.814, 102957...
2    MULTIPOLYGON (((1021176.479 151374.797, 102100...
3    MULTIPOLYGON (((981219.056 188655.316, 980940....
4    MULTIPOLYGON (((1012821.806 229228.265, 101278...
Name: geometry, dtype: geometry
```

(same happens in case of `__getitem__` when `_geometry_column_name` of `gdf3` is None)

**Comment by m-richards** [comment]
> Yes, agreed. Especially now that we have better error messages for this, it might be a good time to reconsider that.

From my perspective it seems like having `_geometry_column_name` defaulting to None is sensible (but keeping keeping the same behaviour that if `geometry` isn't specified in the constructor and a column called `"geometry"` is present then that is detected and set as `_geometry_column_name`.

>I was also wondering, if the geom2 and geometry names are switched (geom2 is active) and then drop geom2, should .geometry then raise the error? 

I think this is clearly confusing behaviour - I've noticed this before as well. Taking a closer look, we actually do raise an attribute error in this case, which is ignored(?), presumably consumed by pandas falls back to the `__getattr__` logic which looks for a column named "geometry"?

Edit:( https://github.com/pandas-dev/pandas/blob/1ab02e474edd7542a59ddbd3f9104dd94a7aec23/pandas/core/generic.py#L5840-L5854
This doesn't treat `geometry` specially so it returns the column based lookup. This doesn't happen generally because if `_geometry_column_name` is present, we look it up directly using `__getitem__`)


So this seems very much unintentional, potentially also a regression? will have a further look.

In any case, I think it should raise an error - by analogy with below
```python
In [29]: gdf = gpd.read_file(gpd.datasets.get_path("nybb"))
In [30]: gdf['columns'] = 4
In [31]: gdf.columns
Out[31]:
Index(['BoroCode', 'BoroName', 'Shape_Leng', 'Shape_Area', 'geometry',
       'columns'],
      dtype='object')
```
 accessing a property of a dataframe in pandas doesn't do a name lookup and it seems reasonable to do the same.

**Comment by m-richards** [comment]
@martinfleis is it possible to move this back to an issue at some point? Maybe I don't have full context on the rationale, I missed the last community meeting and there's not an awful lot on this in the meeting notes. But to me this is actually something actionable as an API choice (but it was tagged as a question so I fully understand why it got moved).

**Reply by martinfleis** [reply]
Ah, sorry! I did a bulk transfer based on a tag and didn't expect anything actionable there, my bad. We cannot just revert the transition to discussions but can create an issue based on a discussion. Let's try to find a consensus here and move it then, the issue shows up only a single post...

**Comment by m-richards** [comment]
No worries, I figured as much, thanks for clarifying
________________________________

**Comment by jorisvandenbossche** [comment]
Hmm, it seems we can't reopen the original issue if we inadvertently do such a "move to discussion"
